### ISSUE TITLE
amp-sidebar: Incorrect focus order/handling

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31168]
## WCAG Level
Level A

## Priority
Low

## Type
Navigation

## Pages/screens/Components affected
* [amp-sidebar](https://amp.dev/documentation/components/amp-sidebar/)

## Description
For people who navigate using a keyboard or keyboard-like device (such as a switch control or a joystick), the focus order of interactive controls in a document should be logical and intuitive.

When a page changes dynamically as a result of user interaction, it's important that focus is, where necessary, moved to the most appropriate place within the page.

`amp-sidebar` looks and behaves like a modal (albeit one that opens to the side, rather than in the center of the viewport). When the sidebar is opened, focus is programmatically moved to the first focusable element in the sidebar.

![Playground example of the sidebar, showing the 'Desktop sidebar' - when opened, focus is on the 'Close sidebar' control which comes after the sidebar's main heading](https://user-images.githubusercontent.com/54933393/99181684-e4288280-2727-11eb-8869-38c6c22dfd58.png)

While this behaviour is appropriate for a strictly marked-up modal dialog that primarily acts as a form or alert, `amp-sidebar` can be used for a variety of content. In these cases, the more appropriate place for the initial focus is the start of the sidebar itself. See also this discussion on [WAI-ARIA Practices - Focus advice for content-heavy dialogs](https://github.com/w3c/aria-practices/issues/442).

## User impact
If the `amp-sidebar` has content that precedes the first focusable element - and, more generally, if the sidebar contains any sort of structured content that goes beyond just a series of buttons or links - assistive technology users may not be aware of this content's presence when the sidebar is opened. They would have to navigate in reverse, from the first focusable element, in order to reach this content. As a result, they may miss this content entirely.

## Required solution
Make sure that the order in which elements receive keyboard focus follows a meaningful and logical sequence. The order does not necessarily have to follow the order implied by the visual layout, as long as it is logical. Nonetheless, the focus should ideally not "jump" erratically between different parts of the layout.

Pay attention to source order. Sequencing elements in the code so they follow a logical order helps to provide a logical and meaningful focus order.

In this case, we would recommend programmatically placing the initial focus to the start of the `amp-sidebar` itself, rather than on the first focusable element. For instance, in the case of the [`amp-sidebar` playground example](https://playground.amp.dev/?url=https%3A%2F%2Fpreview.amp.dev%2Fdocumentation%2Fexamples%2Fcomponents%2Famp-sidebar&format=websites), set programmatic focus to the heading at the start of the sidebar.

To set focus to a static element, which does not ordinarily receive focus, add `tabindex="-1"` to the element - this will then allow the element to be focused using the standard `.focus()` JavaScript method. In the case of the `amp-sidebar` component, we'd recommend adding the `tabindex` attribute dynamically, as part of the sidebar's scripting (rather than expecting authors to add this themselves).

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Turn on your screen reader.
2. Navigate to, and trigger, the sidebar.
3. Verify that focus was placed at the start of the sidebar - the first element in the sidebar should be read out/announced, and using reading keys in the screen reader should allow you to move forward through the document and read through any of the content of the sidebar that precedes the first focusable element.

## Related standards
* [WCAG 2.1 Success Criterion 2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order)

## More information
* [W3C - Placing the interactive elements in an order that follows sequences and relationships within the content](https://www.w3.org/WAI/WCAG21/Techniques/general/G59)
* [The A11Y Project - How-to: Hide content](https://www.a11yproject.com/posts/2013-01-11-how-to-hide-content/)

## Test data
Test date: November 2020
